Migrate Hikvision integration to config flow

ptarjan opened the pull request:
Migrate the Hikvision integration from a legacy YAML-based integration to the modern config flow-based system. In the process of this migration I was able to fix the creation of my binary sensors. Now I have:

<img width="536" height="524" alt="image" src="https://github.com/user-attachments/assets/c4b20458-ca00-4370-856b-e91c04a8473e" />


<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

Users with existing YAML configuration will see a deprecation message.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

- Add config_flow.py for UI-based configuration
- Add const.py with domain and platform constants
- Modernize binary_sensor.py with:
  - Entity names using has_entity_name and translations
  - Device registry support with DeviceInfo
  - Proper callback registration in async_added_to_hass
  - PARALLEL_UPDATES set for push-based updates
- Add strings.json for config flow and entity translations
- Add quality_scale.yaml targeting Bronze tier
- Update manifest.json with config_flow, integration_type, and quality_scale
- Add comprehensive test coverage for:
  - Config flow (user setup, reauth, error handling)
  - Binary sensors (creation, attributes, device classes)
  - Integration setup/unload
- Remove hikvision from INTEGRATIONS_WITHOUT_QUALITY_SCALE_FILE

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [x] I understand the code I am submitting and can explain how it works.
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.
- [x] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

mik-laj commented on homeassistant/components/hikvision/helpers.py:
```diff
@@ -0,0 +1,178 @@
+    response = None
+    for url in urls:
+        try:
+            response = session.get(url, timeout=10)
```
In another PR.

ptarjan replied:
Sounds good. I'll wait for this to land and open another one with that change. Maybe once the upstream PR is merged.

ptarjan committed:
Migrate Hikvision integration to config flow with NVR support

- Add config flow for UI-based setup
- Add reauth flow for credential updates
- Convert to modern entity patterns (has_entity_name, DeviceInfo)
- Add NVR event detection with extended notification methods
- Store runtime data in ConfigEntry.runtime_data
- Add comprehensive tests for config flow, entities, and setup

ptarjan committed:
Address reviewer feedback: use _attr_name for entity naming

Change entity naming from translation_key with placeholders to direct
_attr_name assignment. For cameras, the entity name is just the sensor
type (e.g., "Motion"). For NVRs, the entity name includes the channel
number (e.g., "Motion 1").

This simplifies the implementation and avoids complex translation
placeholder handling.

ptarjan committed:
Add YAML import flow for backward compatibility

Add async_step_import to config_flow.py and async_setup_platform to
binary_sensor.py to support migrating existing YAML configurations
to config entries.

When YAML configuration is detected:
- Automatically import to config entry
- Create deprecation issue in repairs dashboard
- If import fails, create a different issue explaining the failure

This allows users with existing YAML configs to seamlessly transition
to the new config flow-based setup.

ptarjan committed:
Add tests for YAML import flow

Add comprehensive tests for the YAML import functionality:
- test_import_flow: successful import creates config entry
- test_import_flow_with_defaults: uses default port/SSL values
- test_import_flow_cannot_connect: aborts on connection errors
- test_import_flow_no_device_id: aborts when device ID unavailable
- test_import_flow_already_configured: aborts when already configured
- test_yaml_import_creates_deprecation_issue: platform creates issue
- test_yaml_import_creates_failure_issue_on_error: creates failure issue
- test_yaml_import_no_issue_on_already_configured: handles existing config

ptarjan committed:
Remove NVR helper function calls from setup

Remove calls to get_nvr_events and inject_events_into_camera from
async_setup_entry. Also remove the unused fixtures from conftest.py.

ptarjan committed:
Remove unused helpers.py file

ptarjan committed:
Apply suggestion from @mik-laj

Co-authored-by: Kamil Breguła <mik-laj@users.noreply.github.com>

ptarjan committed:
Fix ruff import sorting and remove unused import

ptarjan committed:
Update requirements files

ptarjan committed:
Fix get_id to be method call instead of property

ptarjan committed:
Update generated files from hassfest

ptarjan committed:
python3 -m script.hassfest

ptarjan committed:
Address reviewer feedback: move device class mapping to platform module

Move DEVICE_CLASS_MAP from const.py to binary_sensor.py per reviewer
feedback to keep platform-specific details in the platform module.

ptarjan committed:
Remove defusedxml from requirements

Remove defusedxml as it is not directly used in the integration.
It is a transitive dependency of pyHik.

joostlek commented on homeassistant/components/hikvision/__init__.py:
```diff
@@ -1 +1,86 @@
+    except Exception as err:
+        raise ConfigEntryNotReady(f"Unable to connect to {host}") from err
+
+    device_id = camera.get_id()
```
is this a blocking operation?

joostlek commented on homeassistant/components/hikvision/binary_sensor.py:
```diff
@@ -67,235 +59,175 @@
-    def get_attributes(self, sensor, channel):
-        """Return attribute list for sensor/channel."""
-        return self.camdata.fetch_attributes(sensor, channel)
+    async_add_entities(entities)
```
This can become a list comprehension

joostlek commented on homeassistant/components/hikvision/binary_sensor.py:
```diff
@@ -67,235 +59,175 @@
+    ) -> None:
+        """Initialize the binary sensor."""
+        self._entry = entry
+        self._data = entry.runtime_data
```
we never use the entry

joostlek commented on homeassistant/components/hikvision/config_flow.py:
```diff
@@ -0,0 +1,189 @@
+                )
+                device_id = camera.get_id()
+                device_name = camera.get_name
+            except Exception:
```
Can we be more specific?

joostlek commented on homeassistant/components/hikvision/config_flow.py:
```diff
@@ -0,0 +1,189 @@
+                errors["base"] = "cannot_connect"
+            else:
+                if device_id is None:
+                    errors["base"] = "cannot_connect"
```
Are there devices without device_id?

ptarjan replied:
Yes, device_id can be None. Looking at pyHik's initialize() method:

```
def initialize(self):
    device_info = self.get_device_info()

    if device_info is None:
        self.name = None
        self.cam_id = None  # device_id becomes None
        self.event_states = None
        return
```

This happens when:

* Network errors during get_device_info() (connection timeout, refused, etc.)
* Authentication failures (401 response)
* Device returns invalid/unparseable XML response

pyHik doesn't raise exceptions for these failures - it silently sets `cam_id = None`. Since we're catching `requests.exceptions.RequestException` around the HikCamera() constructor, some errors may still slip through and result in a None device_id.

The check acts as a safety net for cases where pyHik fails to communicate properly but doesn't raise an exception.

ptarjan committed:
Use specific exception type in __init__.py

Use requests.exceptions.RequestException instead of bare Exception
for more precise error handling when connecting to Hikvision devices.

ptarjan committed:
Restore original PLATFORM_SCHEMA for YAML compatibility

Keep the platform schema the same as the original to maintain
backward compatibility with existing YAML configurations.
Adds back CONF_NAME and CONF_CUSTOMIZE fields.

ptarjan committed:
Convert entity creation to list comprehension

Simplify async_setup_entry by using a generator expression
instead of building a list with a loop.

ptarjan committed:
Remove unused self._entry in HikvisionBinarySensor

The entry object is only used to access runtime_data, so we don't
need to store the entry itself.

ptarjan committed:
Use specific exception type in config_flow.py

Use requests.exceptions.RequestException instead of bare Exception
for more precise error handling when connecting to Hikvision devices.

ptarjan committed:
Remove reauth flow for follow-up PR

Remove the reauthentication flow as requested by reviewer.
This can be added in a follow-up PR.

ptarjan committed:
Use CONF_NAME from YAML config for import title

Preserve the user-configured name from YAML configuration when
importing to config entry. Falls back to device name or host.

ptarjan committed:
Use homeassistant domain for deprecated_yaml issue

Use HOMEASSISTANT_DOMAIN for the successful YAML import deprecation
issue, matching the pattern used by other integrations like
nederlandse_spoorwegen. This uses the shared translation strings.

ptarjan committed:
Fix tests for reviewer feedback changes

- Remove reauth flow tests (reauth removed for follow-up PR)
- Use requests.exceptions.RequestException instead of Exception
- Update deprecated_yaml issue tests to use HOMEASSISTANT_DOMAIN

joostlek commented on homeassistant/components/hikvision/binary_sensor.py:
```diff
@@ -67,10 +66,12 @@
+        vol.Optional("ignored", default=False): cv.boolean,
+        vol.Optional("delay", default=0): cv.positive_int,
     }
 )
```
So before we had this customization schema. I think the functionality of ignored can be compared to it being disabled, but what about delay? Is it still an issue today as described in the docs? "Specify the delay to wait after a sensor event ends before notifying Home Assistant in seconds. This is useful to catch multiple quick trips in one window without the state toggling on and off"

joostlek commented on tests/components/hikvision/conftest.py:
```diff
@@ -0,0 +1,102 @@
+        camera.add_update_callback = MagicMock()
+        camera.fetch_attributes = MagicMock(
+            return_value=(False, None, None, "2024-01-01T00:00:00Z")
+        )
```
```suggestion
        camera.fetch_attributes.return_value=(False, None, None, "2024-01-01T00:00:00Z")
```

joostlek commented on tests/components/hikvision/test_binary_sensor.py:
```diff
@@ -0,0 +1,303 @@
+        "homeassistant.components.hikvision.config_flow.HikCamera",
+        autospec=True,
+    ) as hikcamera_mock:
+        camera = hikcamera_mock.return_value
```
this can use the common fixture

joostlek commented on tests/components/hikvision/test_binary_sensor.py:
```diff
@@ -0,0 +1,303 @@
+        await hass.async_block_till_done()
+
+    # Check that deprecation issue was created in homeassistant domain
+    issue_registry = ir.async_get(hass)
```
the issue registry can be loaded as a fixture

joostlek commented on tests/components/hikvision/test_config_flow.py:
```diff
@@ -0,0 +1,348 @@
+            CONF_SSL: False,
+        },
+    )
+    await hass.async_block_till_done()
```
This isn't a blocking operation, so no need to wait

joostlek commented on tests/components/hikvision/test_config_flow.py:
```diff
@@ -0,0 +1,348 @@
+    }
+
+
+async def test_form_with_ssl(
```
1. Can the interface be hosted on a port different than 80 or 443?
2. Why do we ask both the port and the ssl?

ptarjan replied:
1. I don't know. For my NVR it can't be but maybe because the machine is 10 years old. The functionality was there before so I think I'm just leaving it.
2. I assume since you can run on a non-standard port you can choose to use that port with SSL or not.

joostlek commented on tests/components/hikvision/test_config_flow.py:
```diff
@@ -0,0 +1,348 @@
+    assert result2["reason"] == "already_configured"
+
+
+async def test_recovery_from_error(
```
So for the missing ID
1. This test should be merged with the other one
2. If the ID is never there, why do we only show an error instead of aborting

joostlek commented on tests/components/hikvision/test_config_flow.py:
```diff
@@ -0,0 +1,348 @@
+        CONF_USERNAME: TEST_USERNAME,
+        CONF_PASSWORD: TEST_PASSWORD,
+        CONF_SSL: False,
+    }
```
unique id test

joostlek commented on tests/components/hikvision/test_init.py:
```diff
@@ -0,0 +1,105 @@
+    await hass.async_block_till_done()
+
+    assert mock_config_entry.state is ConfigEntryState.NOT_LOADED
+    mock_hikcamera.return_value.disconnect.assert_called_once()
```
Can be merged with the first test

ptarjan committed:
Remove incorrect coordinator comment

ptarjan committed:
Combine mock_hikcamera fixtures and remove unnecessary MagicMock

- Merge mock_hikcamera and mock_hikcamera_config_flow into single fixture
- Patch both HikCamera locations using combined context manager pattern
- Remove unnecessary MagicMock assignments (autospec handles methods)
- Fix fetch_attributes to use return_value assignment syntax

ptarjan committed:
Improve config flow tests per reviewer feedback

- Use combined mock_hikcamera fixture
- Reuse result variable instead of result2, result3
- Remove unnecessary async_block_till_done() calls
- Add unique_id assertions to verify entry uniqueness
- Merge error tests to include recovery (end in CREATE_ENTRY)

ptarjan committed:
Simplify init tests per reviewer feedback

- Merge test_setup_entry and test_unload_entry into single test
- Remove runtime_data internal assertions
- Remove test_setup_entry_no_device_id (implementation detail)
- Remove test_setup_entry_default_name/type tests

ptarjan committed:
Improve binary sensor tests per reviewer feedback

- Use issue_registry fixture instead of getting from hass
- Use combined mock_hikcamera fixture from conftest
- Consolidate entity tests (unique_id, attributes covered in main test)

ptarjan committed:
Add snapshot tests for binary sensor entities

Use snapshot_platform helper to capture entity registry and state
snapshots for all binary sensor entities.

ptarjan committed:
Restore constants to original locations

- Import CONF_DELAY from homeassistant.const (existing constant)
- Define CONF_IGNORED, DEFAULT_DELAY, DEFAULT_IGNORED locally in
  binary_sensor.py to match original code structure

ptarjan committed:
Achieve 100% test coverage for hikvision integration

- Add test for device_id None case in async_setup_entry
- Add test for import flow when device_id is None
- Add test for YAML import abort creating issue
- Add test for binary sensor update callback
- Remove unused timer cancellation code (dead code from delay feature)

ptarjan commented:
Anything you need from me for this PR?

joostlek reviewed (approved):
Awesome!

Will you also look into becoming codeowner? And maybe send me a message on Discord

MartinHjelmare commented on homeassistant/components/hikvision/config_flow.py:
```diff
@@ -0,0 +1,134 @@
+        self._abort_if_unique_id_configured()
+
+        _LOGGER.warning(
+            "Importing Hikvision config from configuration.yaml for %s", host
```
Why is this a warning?

MartinHjelmare commented on tests/components/hikvision/conftest.py:
```diff
@@ -0,0 +1,94 @@
+
+
+@pytest.fixture
+async def init_integration(
```
This fixture is never used.
